### PR TITLE
moving files from personal fork to awslabs repo

### PR DESCRIPTION
**Pull request by access2rohit**
Single GPU inference for RESNET 50.

### CODE REVIEW

File: utils/cpu_gpu_profiler.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 
 # redirect the GPU memory usage to a file
 #GPU_MONITOR = "nvidia-smi --query-gpu=index,memory.used --format=csv -lms 500 -f output.csv"
-GPU_MONITOR = "nvidia-smi --query-gpu=index,memory.used --format=csv -l 120 -f output.csv"
+GPU_MONITOR = "nvidia-smi --query-gpu=index,memory.used --format=csv -l 50 -f output.csv"
```
Comment (piyushghai) [inline utils/cpu_gpu_profiler.py]: Should this value change ?

Comment (access2rohit) [reply utils/cpu_gpu_profiler.py]: This experiment finishes in less than 70 secs so I needed to sample nvidia-smi dump before that 50 seems like a safe range to atleast get one sample. Is there an alternate way to configure nvidia-smi dump time for a specific banchmark ?

Comment (piyushghai) [reply utils/cpu_gpu_profiler.py]: Maybe @leleamol might have an idea here.

Comment (leleamol) [reply utils/cpu_gpu_profiler.py]: @access2rohit ,
The file is used for all the jobs that are running in BAI today.  Decreasing the time may impact performance of other tests as nvidia-smi freezes the GPUs when the measurements are taken. Currently, there isn't a way to set the sampling time for an individual run.  Is it possible to increase the run time of your experiment or running it multiple times and then sampling the memory usage?

**Comment by access2rohit** [comment]
@leleamol I have updated the code to run more than 120 secs and hence no longer require to dump nvidia-smi in 50 secs. Can you review ?

**Comment by access2rohit** [comment]
These changes have been tested manually on single V100 instance. Once approved I will make changes to their respective cloudwatch rules to make sure they are picked from here and then monitor then for a week.

File: mxnet_benchmark/train_imagenet.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
+import argparse, time, logging, os
```
Comment (apeforest) [inline mxnet_benchmark/train_imagenet.py]: If you are adding a BAI test for inference in symbolic mode, why is this needed?

File: mxnet_benchmark/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
+import logging
```
Comment (apeforest) [inline mxnet_benchmark/lr_scheduler.py]: Same question, why is this file needed? called anywhere?

File: mxnet_benchmark/infer_imagenet.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
+import argparse, time, logging, os
```
Comment (larroy) [inline mxnet_benchmark/infer_imagenet.py]: can we add (python3) shebang?

File: mxnet_benchmark/infer_imagenet.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+import numpy as np
+import mxnet as mx
+from mxnet import gluon, nd
+from mxnet.gluon.data.vision import transforms
+from gluoncv.data import imagenet
+from gluoncv.model_zoo import get_model
+from gluoncv.utils import makedirs
+
+# CLI
```
Comment (larroy) [inline mxnet_benchmark/infer_imagenet.py]: Would be nicer to put in main or separate function for argument parsing.

File: mxnet_benchmark/infer_imagenet.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+context = [mx.gpu(int(i)) for i in opt.num_gpus.split(',')] if num_gpus > 0 else [mx.cpu()]
+num_workers = opt.num_workers
+
+kv = mx.kv.create(opt.kvstore)
+model_name = opt.model
+
+kwargs = {'ctx': context, 'pretrained': opt.use_pretrained, 'classes': classes}
+net = get_model(model_name, **kwargs)
+
+def get_data_rec(rec_val, rec_val_idx, batch_size, num_workers):
```
Comment (larroy) [inline mxnet_benchmark/infer_imagenet.py]: docstring missing, return type would be nice

File: mxnet_benchmark/infer_imagenet.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+if opt.use_rec:
+    val_data, batch_fn = get_data_rec(opt.rec_val, opt.rec_val_idx, batch_size, num_workers)
+else:
+    val_data, batch_fn = get_data_loader(opt.data_dir, batch_size, num_workers)
+
+acc_top1 = mx.metric.Accuracy()
+acc_top5 = mx.metric.TopKAccuracy(5)
+
+def infer(ctx, count):
```
Comment (larroy) [inline mxnet_benchmark/infer_imagenet.py]: docstring, same as above.

File: mxnet_benchmark/lr_scheduler.py (hunk 3 above)
Comment (access2rohit) [reply mxnet_benchmark/lr_scheduler.py]: @apeforest These files  are being added since there are other functions in lr_scheduler and they are used in training models. Although this commit will currently be only used for inference purposes but the intention of adding training and scheduler files is to make addition of a new training BAI of resnet50_v2 easier(if required in future).

File: mxnet_benchmark/infer_imagenet.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        if i<20:
+            logging.info('warmup_throughput: %d samples/sec warmup_time %f'%(
+                     int(batch_size / time_taken), time_taken))
+        else:
+            if count>0:
+                logging.info('Speed: %d samples/sec Time cost=%f'%(
+                         int(batch_size / time_taken), time_taken))
+    return
+
+def main():
```
Comment (larroy) [inline mxnet_benchmark/infer_imagenet.py]: Suggestion for handling returns

```

def main():

    return 1

if __name__ == '__main__':
    sys.exit(main())

```

File: mxnet_benchmark/lr_scheduler.py (hunk 3 above)
Comment (larroy) [reply mxnet_benchmark/lr_scheduler.py]: standard python logging

File: mxnet_benchmark/train_imagenet.py (hunk 2 above)
Comment (access2rohit) [reply mxnet_benchmark/train_imagenet.py]: please refer to above answer

File: mxnet_benchmark/train_imagenet.py [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+# pip install gluoncv for model definitions
+
+from gluoncv.data import imagenet
+from gluoncv.model_zoo import get_model
+from gluoncv.utils import makedirs
+import re, math
+from lr_scheduler import MultiFactorScheduler, PolyScheduler, WarmupScheduler
+
+# CLI
+parser = argparse.ArgumentParser(description='Train a model for image classification.')
```
Comment (larroy) [inline mxnet_benchmark/train_imagenet.py]: Same comments as above, shebang and parsing in main or other.

File: mxnet_benchmark/train_imagenet.py [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+                    help='frequency of model saving.')
+parser.add_argument('--save-dir', type=str, default='params',
+                    help='directory of saved models')
+parser.add_argument('--logging-dir', type=str, default='logs',
+                    help='directory of training logs')
+parser.add_argument('--kvstore', type=str, default='nccl')
+parser.add_argument('--top-k', type=int, default=0, help='give 5 for top5 accuracy, if 0 only prints top1 accuracy')
+opt = parser.parse_args()
+
+logging.basicConfig(level=logging.INFO)
```
Comment (larroy) [inline mxnet_benchmark/train_imagenet.py]: Suggest to configure logging on the top of main.

File: mxnet_benchmark/lr_scheduler.py (hunk 3 above)
Comment (apeforest) [reply mxnet_benchmark/lr_scheduler.py]: I would suggest to add this later when you add the training BAI test. Keep each PR simple and dedicated.

File: mxnet_benchmark/train_imagenet.py [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+    return (1-top1, 1-top5)
+
+def train(ctx):
+    if isinstance(ctx, mx.Context):
+        ctx = [ctx]
+    net.initialize(initializer, ctx=ctx)
+
+    trainer = gluon.Trainer(net.collect_params(), optimizer, optimizer_params, kvstore=kv)
+
+    L = gluon.loss.SoftmaxCrossEntropyLoss()
```
Comment (larroy) [inline mxnet_benchmark/train_imagenet.py]: can we use something better than one letter variable names?

File: mxnet_benchmark/train_imagenet.py [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+
+if opt.lr_mode == 'poly':
+    lr_scheduler = PolyScheduler(opt.num_epochs * epoch_size, base_lr=opt.lr, pwr=opt.lr_poly_power, warmup_steps=(opt.warmup_epochs * epoch_size))
+elif opt.lr_mode == 'step':
+    lr_decay = opt.lr_decay
+    lr_decay_epoch = [int(i) for i in opt.lr_decay_epoch.split(',')]
+    steps = [epoch_size * x for x in lr_decay_epoch]
+    lr_scheduler = MultiFactorScheduler(step=steps, factor=lr_decay, base_lr=opt.lr)
+else:
+    raise ValueError
```
Comment (larroy) [inline mxnet_benchmark/train_imagenet.py]: suggest to add a string message to exception

File: mxnet_benchmark/lr_scheduler.py (hunk 3 above)
Comment (access2rohit) [reply mxnet_benchmark/lr_scheduler.py]: ok

File: mxnet_benchmark/train_imagenet.py [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+                    help='Number of batches to wait before logging.')
+parser.add_argument('--save-frequency', type=int, default=0,
+                    help='frequency of model saving.')
+parser.add_argument('--save-dir', type=str, default='params',
+                    help='directory of saved models')
+parser.add_argument('--logging-dir', type=str, default='logs',
+                    help='directory of training logs')
+parser.add_argument('--kvstore', type=str, default='nccl')
+parser.add_argument('--top-k', type=int, default=0, help='give 5 for top5 accuracy, if 0 only prints top1 accuracy')
+opt = parser.parse_args()
```
Comment (larroy) [inline mxnet_benchmark/train_imagenet.py]: usually is named "args" as in python documentation, would be nice to be consistent across our scripts.

File: mxnet_benchmark/train_imagenet.py [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+from gluoncv.model_zoo import get_model
+from gluoncv.utils import makedirs
+import re, math
+from lr_scheduler import MultiFactorScheduler, PolyScheduler, WarmupScheduler
+
+# CLI
+parser = argparse.ArgumentParser(description='Train a model for image classification.')
+parser.add_argument('--data-dir', type=str, default='~/.mxnet/datasets/imagenet',
+                    help='training and validation pictures to use.')
+parser.add_argument('--rec-train', type=str, default='/media/ramdisk/data/train-passthrough.rec',
```
Comment (larroy) [inline mxnet_benchmark/train_imagenet.py]: Can we add a note about these hardcoded paths? where are they coming from?

File: mxnet_benchmark/infer_imagenet.py (hunk 4 above)
Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: These runs are supposed to be python version independent. Could you explain why should we add version specific python shebang ?

File: mxnet_benchmark/infer_imagenet.py (hunk 6 above)
Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: @larroy could you elaborate what are you suggesting here ?

File: mxnet_benchmark/infer_imagenet.py (hunk 7 above)
Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: same as above

File: mxnet_benchmark/infer_imagenet.py (hunk 5 above)
Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: good suggestion !

File: mxnet_benchmark/infer_imagenet.py (hunk 6 above)
Comment (larroy) [reply mxnet_benchmark/infer_imagenet.py]: I believe is a good practice. If it's py2 py3 compatible:

```
#!/usr/bin/env python
```
Is a good choice.

It won't work if called as an executable script without shebang for example.

Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: its always called as `python <script name>` but if 
`#!/usr/bin/env python` works then that sounds good

File: mxnet_benchmark/infer_imagenet.py (hunk 8 above)
Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: addressed

**Comment by access2rohit** [comment]
@larroy  @apeforest  can you review

File: mxnet_benchmark/infer_imagenet.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+import numpy as np
+import mxnet as mx
+from mxnet import gluon, nd
+from mxnet.gluon.data.vision import transforms
+from gluoncv.data import imagenet
+from gluoncv.model_zoo import get_model
+from gluoncv.utils import makedirs
+
+
+parser = argparse.ArgumentParser(description='Train a model for image classification.')
```
Comment (larroy) [inline mxnet_benchmark/infer_imagenet.py]: Why not do all this parsing in main? instead of global scope?

**Comment by access2rohit** [comment]
@mseth10: I didn't know that. Thanks for pointing that out !

**Comment by access2rohit** [comment]
@apeforest @mseth10  @larroy @anirudh2290 @leleamol addressed comments and updated PR. please review

File: mxnet_benchmark/infer_imagenet.py (hunk 8 above)
Comment (access2rohit) [reply mxnet_benchmark/infer_imagenet.py]: I don't think this is good idea in this case because the intention is to throw an error in BAI as the step function hardly gives any exception to help debug the error. I will be keeping main function call as it is for now

File: image_classification/infer_imagenet_gpu.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+def get_data_rec(rec_val, rec_val_idx, batch_size, num_workers):
+    """
+       Creates and returns data MXNet Data Iterator object and a function that splits data into batches
+       (if using image record iter for input)
+    """
+    rec_val = os.path.expanduser(rec_val)
+    rec_val_idx = os.path.expanduser(rec_val_idx)
+    mean_rgb = [123.68, 116.779, 103.939]
+    
+    def batch_fn(batch, ctx):
```
Comment (larroy) [inline image_classification/infer_imagenet_gpu.py]: Could we find a better name for this function? Like split_data_label_fn ?  batch_fn doesn't say much.

File: image_classification/infer_imagenet_gpu.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+        outputs = [net(X.astype(opt.dtype, copy=False)) for X in data]
+        acc_top1.update(label, outputs)
+        logging.info('Batch [%d]'%(i))
+        logging.info('Top 1 accuracy: %d'%(acc_top1.get()[1]))
+        time_taken = time.time() - btic
+        if i<20:
+            logging.info('warmup_throughput: %d samples/sec warmup_time %f'%(
+                     int(batch_size / time_taken), time_taken))
+        else:
+
```
Comment (larroy) [inline image_classification/infer_imagenet_gpu.py]: can we remove this empty line?

File: image_classification/infer_imagenet_gpu.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+                        help='the index of validation data')
+    parser.add_argument('--use-rec', action='store_true',
+                        help='use image record iter for data input. default is false.')
+    parser.add_argument('--batch-size', type=int, default=32,
+                        help='training batch size per device (CPU/GPU).')
+    parser.add_argument('--dataset-size', type=int, default=32,
+                        help='training batch size per device (CPU/GPU).')
+    parser.add_argument('--dtype', type=str, default='float32',
+                        help='data type for training. default is float32')
+    parser.add_argument('--num_gpus', type=str, default=0,
```
Comment (larroy) [inline image_classification/infer_imagenet_gpu.py]: Im a bit confused, is the argument a number? below there's a split with a comma, does one have to list the gpus?

File: image_classification/infer_imagenet_gpu.py [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+    return
+
+def main():
+    opt = arg_parser()
+    logging.basicConfig(level=logging.INFO)
+    logging.info(opt)
+ 
+    batch_size = opt.batch_size
+    classes = 1000
+    num_gpus = len(opt.num_gpus.split(','))
```
Comment (larroy) [inline image_classification/infer_imagenet_gpu.py]: Is it a list or a number? see above

File: image_classification/infer_imagenet_gpu.py [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+        data = gluon.utils.split_and_load(batch.data[0], ctx_list=ctx, batch_axis=0)
+        label = gluon.utils.split_and_load(batch.label[0], ctx_list=ctx, batch_axis=0)
+        return data, label
+    
+    val_data = mx.io.ImageRecordIter(
+        path_imgrec         = rec_val,
+        path_imgidx         = rec_val_idx,
+        preprocess_threads  = num_workers,
+        shuffle             = False,
+        batch_size          = batch_size,
```
Comment (anirudh2290) [inline image_classification/infer_imagenet_gpu.py]: do we run inference on real dataset or dummy dataset for the benchmark ?

File: image_classification/infer_imagenet_gpu.py [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+    model_name = opt.model
+    kwargs = {'ctx': context, 'pretrained': opt.use_pretrained, 'classes': classes}
+
+    if opt.use_rec:
+        val_data, batch_fn = get_data_rec(opt.rec_val, opt.rec_val_idx, batch_size, num_workers)
+    else:
+        val_data, batch_fn = get_data_loader(opt.data_dir, batch_size, num_workers, opt)
+    
+    acc_top1 = mx.metric.Accuracy()
+    acc_top5 = mx.metric.TopKAccuracy(5)
```
Comment (anirudh2290) [inline image_classification/infer_imagenet_gpu.py]: does this accuracy actually work. Do we measure accuracy for this benchmark ?

File: image_classification/infer_imagenet_gpu.py [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+        if opt.dtype == 'float16':
+            data = mx.sym.Cast(data=data, dtype=np.float16)
+            net.cast(np.float16)
+        out = net(data)
+        if opt.dtype == 'float16':
+            out = mx.sym.Cast(data=out, dtype=np.float32)
+        softmax = mx.sym.SoftmaxOutput(out, name='softmax')
+        mod = mx.mod.Module(softmax, context=context)
+        net.hybridize()
+        if opt.dtype == 'float16':
```
Comment (anirudh2290) [inline image_classification/infer_imagenet_gpu.py]: I couldnt see the dashboard for FP16 for inference ? Is there a dashboard ?

File: image_classification/infer_imagenet_gpu.py (hunk 21 above)
Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: It does but we don't care about this during inference

File: image_classification/infer_imagenet_gpu.py (hunk 22 above)
Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: someone removed it from dashboard after accesses were revoked. But there are cloudwatch rules for the same and data is still getting collected

Comment (mseth10) [reply image_classification/infer_imagenet_gpu.py]: You will have to get edit access to benchmark account to modify your benchmark file location. Can you please add a dashboard then?

File: image_classification/infer_imagenet_gpu.py (hunk 20 above)
Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: It is using real data. Refer to lines 17 and 19, which are used in lines 92 and 93.

Comment (anirudh2290) [reply image_classification/infer_imagenet_gpu.py]: what are the args passed for --rec-val-idx and --rec-val

File: image_classification/infer_imagenet_gpu.py (hunk 21 above)
Comment (anirudh2290) [reply image_classification/infer_imagenet_gpu.py]: I guess this is future work but for running inference for pretrained fp32 models on fp16, accuracy is important to measure.  When I run this script for inference with fp16 the accuracy always shows all 0s. Do you know why ?

Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: The inference benchmarks do not show good accuracy. I don't know why. Existing BAI Inference dashboards report latency or speed along with memory consumption only. Do you think accuracy matters in case of inference ?

File: image_classification/infer_imagenet_gpu.py (hunk 20 above)
Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: command to be executed `python image_classification/infer_imagenet_gpu.py --use-rec --batch-size 128 --dtype float32 --num-data-workers 40 --mode symbolic --model resnet50_v2 --num_gpus 0 --use-pretrained --kvstore local --log-interval 1 --num-epochs 1 --rec-val ~/data/val-passthrough.rec --rec-val-idx ~/data/val-passthrough.idx`

Comment (anirudh2290) [reply image_classification/infer_imagenet_gpu.py]: is "passthrough" real data or dummy data ?

Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: Its real. Please refer to the links to S3 bucket and the code that downloads it:

https://s3.console.aws.amazon.com/s3/buckets/aws-ml-platform-datasets/imagenet/pass-through/?region=us-east-1&tab=overview

https://github.com/access2rohit/deeplearning-benchmark/blob/6984abb4cc08aa291e284835b2dc2a7229c05abf/utils/data_manager.py#L22

File: image_classification/infer_imagenet_gpu.py (hunk 21 above)
Comment (anirudh2290) [reply image_classification/infer_imagenet_gpu.py]: as i said it may be useful for fp16 case but not blocking.

File: image_classification/infer_imagenet_gpu.py (hunk 20 above)
Comment (anirudh2290) [reply image_classification/infer_imagenet_gpu.py]: cool thanks

File: image_classification/infer_imagenet_gpu.py (hunk 22 above)
Comment (access2rohit) [reply image_classification/infer_imagenet_gpu.py]: sure
